File: 021/MergeTwoSortedLists.scala
/**
 * Definition for singly-linked list.
 * class ListNode(var _x: Int = 0) {
 *   var next: ListNode = null
 *   var x: Int = _x
 * }
 */

class ListNode(var _x: Int = 0) {
    var next: ListNode = null
    var x: Int = _x
}

object MergeTwoSortedLists {
    def mergeTwoLists(l1: ListNode, l2: ListNode): ListNode = {
        val dummyHead = new ListNode(0)
        var p = dummyHead
        var lr1 = l1
        var lr2 = l2
        while (lr1 != null && lr2 != null) {
            if (lr1._x < lr2._x) {
                p.next = lr1
                lr1 = lr1.next
            } else {
                p.next = lr2
                lr2 = lr2.next
            }
            p = p.next
        }
        
        if (lr1 != null)
            p.next = lr1
        if (lr2 != null)
            p.next = lr2
        dummyHead.next
    }
}

File: 021/MergeTwoSortedLists.kt
/**
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int = 0) {
 *     var next: ListNode? = null
 * }
 */
class MergeTwoSortedLists {
    fun mergeTwoLists(l1: ListNode?, l2: ListNode?): ListNode? {
        if (l1 == null)
            return l2
        if (l2 == null)
            return l1
        val re: ListNode
        if (l1.`val` < l2.`val`) {
            re = l1
            re.next = mergeTwoLists(l1.next, l2)
        } else {
            re = l2
            re.next = mergeTwoLists(l1, l2.next)
        }
        return re
    }
}